### DISCUSSION TITLE
Unable to save image as lossy jpeg2000

### DISCUSSION THREAD

**Discussion by pimlottc** [ImageMagick/ImageMagick#6672] [Contribute]
### ImageMagick version

7.1.1-5

### Operating system

MacOS

### Operating system, version and so on

macOS 13.3 (Ventura)

### Description

I can't seem to get ImageMagick to create a lossy jpeg2000.  Jpeg2000 supports both lossless and lossy compression, but ImageMagick is always using lossless.

The [jpeg2000 format docs](http://patrickdieudonne.com/ImageMagick-6.7.0-2/www/jp2.html) are not explicit about how to choose lossy vs lossless, but I *think* from the first example shown that it's supposed to happen automatically based on the value of `jp2:rate` encoding option:

> * compression is lossy at 64:1
> `-define jp2:rate=0.015625`

However, when I try this same command, the output file is reported to be lossless (using ffprobe).

### Steps to Reproduce

```
% convert wizard.png -define jp2:cblkwidth=64 -define jp2:cblkheight=32 \
-define jp2:nomct -define jp2:numrlvls=4 -define jp2:rate=0.015625 wizard.jpc
```

```
% ffprobe wizard.jpc 2>&1 | grep "Stream"
  Stream #0:0: Video: jpeg2000, rgb24, 940x1250, lossless, 25 fps, 25 tbr, 25 tbn
```

### Images

![wizard](https://user-images.githubusercontent.com/188955/229563904-90a3306f-3583-475b-9b22-711e10a59966.png)
[wizard.jpc.zip](https://github.com/ImageMagick/ImageMagick/files/11140172/wizard.jpc.zip)

**Comment by pimlottc** [comment]
Additionally, if I generate two jp2 files using the example command but with different rates, the output files sizes are identical.

```
% convert wizard.png -define jp2:cblkwidth=64 -define jp2:cblkheight=32 \
-define jp2:nomct -define jp2:numrlvls=4 -define jp2:rate=1 wizard.1-1.jp2
% convert wizard.png -define jp2:cblkwidth=64 -define jp2:cblkheight=32 \
-define jp2:nomct -define jp2:numrlvls=4 -define jp2:rate=0.015625 wizard.64-1.jp2
% ls -la wizard.*.jp2
634707 Apr  3 11:02 wizard.1-1.jp2
634707 Apr  3 11:02 wizard.64-1.jp2
```

They also appear visually identical (i.e. no compression artifacts).

[wizard.rates.zip](https://github.com/ImageMagick/ImageMagick/files/11140253/wizard.rates.zip)

**Comment by pimlottc** [comment]
Looking a the OpenJPEG docs, it looks like the [irreversible parameter](http://www.openjpeg.org/doxygen/structopj__cparameters.html#a47596276043d689a1c18c0c54272c489
) must set to `1` to allow lossy compression (default is `0`).  In the jp2 coder, this only happens when it detects that the image matches the "Digital Cinema 4K profile" in the [CinemaProfileCompliance](https://github.com/ImageMagick/ImageMagick/blob/9a9896fce95d09e5e47b86baccbe1ce1a2fca76b/coders/jp2.c#L737) function.

**Comment by pimlottc** [comment]
Verified output images are all lossless using jpylyzer:

```
wizard.1-1.jp2
                <transformation>5-3 reversible</transformation>
        <compressionRatio>5.55</compressionRatio>
wizard.64-1.jp2
                <transformation>5-3 reversible</transformation>
        <compressionRatio>5.55</compressionRatio>
wizard.jpc
                <transformation>5-3 reversible</transformation>
        <compressionRatio>5.55</compressionRatio>
```

**Comment by dlemstra** [comment]
Did a quick test what would happen if we enable `irreversible` and the size is reduced but I am not sure if the rate is used because I am getting the same size for both images:

```
177885 Apr  3 20:07 a.jp2
177885 Apr  3 20:07 b.jp2
```

So I am not sure when we should enable this flag. Maybe it should be a separate option?

P.s. the following defines do nothing: `-define jp2:cblkwidth=64 -define jp2:cblkheight=32 -define jp2:nomct -define jp2:numrlvls=4`.

**Comment by pimlottc** [comment]
The example/defines are perhaps relics from when ImageMagick was using Jasper instead of OpenJPEG?

**Comment by pimlottc** [comment]
I think 2599725a6c1f299a7d44edd55abf77ca5537b5d8 is the commit that changed to using OpenJPEG

From the [old changelog](https://bricomoral.com/ImageMagick-6.8.9-7/www/changelog.html) (which seems to have been removed in the move to github?) 
> 014-02-21 6.8.8-8 Cristy <quetzlzacatenango@image...>
> - Replace MAGICKCORE_JP2_DELEGATE with MAGICKCORE_LIBOPENJP2_DELEGATE in magick/static.c (reference http://www.imagemagick.org/discourse-server/viewtopic.html?f=3&t=25153).
